### ISSUE TITLE
Magisk is not automatically reinstalled on LineageOS updates since 18.1

### ISSUE THREAD

**Issue by christophehenry** [topjohnwu/Magisk#4655]
Device: OnePlus 5T
Android version: 11 (LineageOS 18.1)
Magisk version name: 23.0
Magisk version code: 23001

Following the suggestion [here](https://github.com/topjohnwu/Magisk/issues/4394#issuecomment-903088503), I'm opening the issue to raise the concern that Magisk is not automatically reinstalled on LineageOS updates since 18.1. 

This issue has persisted accross all Lineage's updates despite the presence of the `99-magisk.sh` in `/system/addon.d/`. I also asked to Lineage's team if this was a known issue and their response was basically "Yeah, we don't care" which is disappointing.

I need Magisk for several app, including automated backups using Titanium and app isolation with AFWall+. I can't afford to run Android without it.

**Comment by gbschenkel** [comment]
This happens because OnePlus 5T doesn't have A/B partition. Each time a update is installed it will overwrite the Magisk modifications. For you be able to upgrade and maintain Magisk you will need patch the OTA. I didn't find a easy way.
I have and use the OnePlus 5 (cheeseburger).

**Comment by christophehenry** [comment]
> I reinstall Magisk after the OTA upgrade, using adb probably as you 

Yes, that's what I do.

I don't understand. People from LineageOS told me that GApps reinstall after updates through `/system/addon.d/` worked correctly. How is it different from Magisk?

**Comment by Didgeridoohan** [comment]
IIRC the discussion on this ended with both LOS devs and John saying: "but I/we want it this way and that doesn't work with your way". Which means that it's unlikely to be fixed, ever...

**Comment by t-aus-m** [comment]
> This happens because OnePlus 5T doesn't have A/B partition.

That might be wrong. I have the same issues on my Fairphone 3, which is a A/B device.

**Comment by topjohnwu** [comment]
@osm0sis would you be able to take a look?

**Comment by osm0sis** [comment]
On my to-do to investigate already, just very little time at the moment. I'll see what I can work on when I'm off in the few weeks!

**Comment by christophehenry** [comment]
If I can be of any help, tell me.

**Comment by ChenDanCMU** [comment]
Same issue here on a Mi MIX (lithium) LOS 18.1.

**Comment by gbschenkel** [comment]
> > This happens because OnePlus 5T doesn't have A/B partition.
> 
> That might be wrong. I have the same issues on my Fairphone 3, which is a A/B device.

Was the response I got at LineageOS channel on Libera.chat

**Comment by jasonl367** [comment]
Same issue. galaxy s10, official los 18.1

**Comment by coxackie** [comment]
Isn't the main problem [this](https://github.com/topjohnwu/Magisk/pull/3037)? It has nothing to do with A/B partitions - it is a problem whenever one uses lineage recovery (that does not support FBE decryption). There is even a [fork](https://github.com/programminghoch10/Lygisk) recently created to deal with the fact gives `/data` access in `addon.d`.

**Comment by osm0sis** [comment]
Yep, might be that if they're using Lineage Recovery, but if they're using TWRP with working decryption then could be something else. Maybe @christophehenry can describe his setup a bit more.

**Comment by Iey4iej3** [comment]
To add: I don't have this issue but I use TWRP and I install the updates manually in the recovery. However, it is unclear if it is due to Lineage Recovery "since 18.1", since the full-disk encryption was applied much earlier.

**Comment by coxackie** [comment]
@osm0sis granted, it could be one or the other; the fact is, there is a large basis of lineage users that faces this same issue due to lineage recovery, and it is a bit sad that they either have to do update manually every time, or now have to use a fork that will try to keep up with upstream. but, oh well, decisions are decisions.

**Comment by cocka11** [comment]
> Device: OnePlus 5T
> Android version: 11 (LineageOS 18.1)
> Magisk version name: 23.0
> Magisk version code: 23001
> 
> Following the suggestion [here](https://github.com/topjohnwu/Magisk/issues/4394#issuecomment-903088503), I'm opening the issue to raise the concern that Magisk is not automatically reinstalled on LineageOS updates since 18.1.
> 
> This issue has persisted accross all Lineage's updates despite the presence of the `99-magisk.sh` in `/system/addon.d/`. I also asked to Lineage's team if this was a known issue and their response was basically "Yeah, we don't care" which is disappointing.
> 
> I need Magisk for several app, including automated backups using Titanium and app isolation with AFWall+. I can't afford to run Android without it.

i am in xiaomi redmi 7 and use lineage recovery and i did not have any this problem during the update. 🤔

**Comment by christophehenry** [comment]
@osm0sis Yes, indeed, during 18.1 update, LineageOS was uncivil enough to replace TWRP by its own recovery. I could try to reflash TWRP. But then I'm afraid Lineage replaces it again on each major release or even minor since, from what i've heard, they decided to stop supporting TWRP?

**Comment by Iey4iej3** [comment]
> @osm0sis Yes, indeed, during 18.1 update, LineageOS was uncivil enough to replace TWRP by its own recovery. I could try to reflash TWRP. But then I'm afraid Lineage replaces it again on each major release or even minor since, from what I've heard, they decided to stop supporting TWRP?

I don't know whether they decided to stop supporting TWRP, but you could disable `System → Advanced → Updater → ⫶ → Preferences → Update recovery`. I don't use OTA updates though.

**Comment by osm0sis** [comment]
If not in the updater I think there's a Developer Setting too for disabling Lineage Recovery flashing.

**Comment by christophehenry** [comment]
This is disappointing for LineageOS but I understand it's not up to you if this is confirmed. At least I have a fix I can test. You might want to add this in the documentation, though.

**Comment by osm0sis** [comment]
Yeah it's just a limitation of Lineage Recovery not supporting decrypting data. There were options/proposals for working around this by storing Magisk binaries in system with the addon.d scripts, but that's a bit ugly to have multiple copies of Magisk files around which could get out of sync, so John was never really interested in that solution.

**Comment by osm0sis** [comment]
Whoops, tapped by mistake. Still want to do the testing I mentioned above. 😅

**Comment by Iey4iej3** [comment]
There was a post which indicated that it works on a Xiaomi device with Lineage Recovery. Seemingly one still needs to look into that.

**Comment by ARitz-Cracker** [comment]
> @osm0sis Yes, indeed, during 18.1 update, LineageOS was uncivil enough to replace TWRP by its own recovery. I could try to reflash TWRP. But then I'm afraid Lineage replaces it again on each major release or even minor since, from what I've heard, they decided to stop supporting TWRP?

Hey there! I'm also running a OnePlus 5t with LineageOS. Just wanted to let you know that you can disable updating recovery in the OTA updates settings, and installing OTA updates using TWRP works fine for now!

Though it'll definitely be interesting if Magisk can survive updates in a way which does not require /data decryption
